src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -68,8 +76,13 @@ public static int NeedsEscaping(ReadOnlySpan<byte> value, JavaScriptEncoder enco
             return idx;
         }
 
-        public static int NeedsEscaping(ReadOnlySpan<char> value)
+        public static int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
+            if (encoder != null)
+            {
+                return encoder.FindFirstCharacterToEncodeUtf8(MemoryMarshal.Cast<char, byte>(value));
```

Comment:
For those who might come across this PR/change in the future, this use of `MemoryMarshal.Cast` to go from char to bytes is not valid since the bytes are still UTF-16 encoded characters (and not UTF-8 encoded text). Transformations like these should be done via `Encoding.UTF8` (and the data actually transcoded). Re-interpreting the chars as bytes and passing them to an API that expects UTF-8 is incorrect and results in bugs around the `Utf8JsonWriter` escaping logic (stackoverflow, characters that should be escaped being skipped, etc.).

**Note:** What's in master and release/3.0 branches has been fixed and doesn't contain the bug (it uses the built-in `JavascriptEncoder.FindFirstCharacterToEncode` method that accepts UTF-16 chars as input and hence avoids having to transcode to UTF-8 all-together).

Here's the issue: https://github.com/dotnet/corefx/issues/40979
Here's the fix: https://github.com/dotnet/corefx/pull/40996 (master) / https://github.com/dotnet/corefx/pull/40997 (release/3.0).